File: mobile-app/smartspace_mobile/lib/services/user_service.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'storage_service.dart';

class UserService {
  final SecureStorageService _storageService = SecureStorageService();

  static String get baseUrl {
    if (kIsWeb) {
      return 'http://localhost:5030/api/users';
    } else if (defaultTargetPlatform == TargetPlatform.android) {
      return 'http://10.0.2.2:5030/api/users';
    } else {
      return 'http://localhost:5030/api/users';
    }
  }

  /// Fetch all users with optional partial search.
  Future<List<Map<String, dynamic>>> fetchUsers({String searchQuery = ''}) async {
    final token = await _storageService.getToken();

    final uri = Uri.parse(
      searchQuery.trim().isNotEmpty
          ? '$baseUrl?search=${Uri.encodeComponent(searchQuery.trim())}'
          : baseUrl,
    );

    final response = await http.get(
      uri,
      headers: {
        'Content-Type': 'application/json',
        if (token != null) 'Authorization': 'Bearer $token',
      },
    ).timeout(const Duration(seconds: 8));

    if (response.statusCode == 200) {
      final List<dynamic> data = jsonDecode(response.body);
      return data.cast<Map<String, dynamic>>();
    } else {
      final data = jsonDecode(response.body).catchError((_) => {});
      throw Exception(data['message'] ?? 'Failed to fetch users (Status ${response.statusCode})');
    }
  }

  /// Update a user's role.
  Future<Map<String, dynamic>> updateUserRole({
    required String id,
    required String newRole,
  }) async {
    final token = await _storageService.getToken();

    final response = await http.put(
      Uri.parse('$baseUrl/$id/role'),
      headers: {
        'Content-Type': 'application/json',
        if (token != null) 'Authorization': 'Bearer $token',
      },
      body: jsonEncode({'newRole': newRole}),
    ).timeout(const Duration(seconds: 8));

    final data = jsonDecode(response.body);

    if (response.statusCode == 200) {
      return data as Map<String, dynamic>;
    } else {
      throw Exception(data['message'] ?? 'Failed to update user role');
    }
  }
}

File: frontend-web/src/services/userService.js
const API_BASE_URL = 'http://localhost:5030/api';

/**
 * Helper to retrieve stored auth token from local storage.
 */
function getAuthToken() {
  return localStorage.getItem('smartspace_token') || localStorage.getItem('token');
}

/**
 * Service to manage users and role assignments.
 */
export const userService = {
  /**
   * Fetch users with optional search query (filters FullName and Email).
   * @param {string} searchQuery
   */
  async fetchUsers(searchQuery = '') {
    const token = getAuthToken();
    const url = new URL(`${API_BASE_URL}/users`);
    if (searchQuery && searchQuery.trim() !== '') {
      url.searchParams.append('search', searchQuery.trim());
    }

    const response = await fetch(url.toString(), {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        ...(token ? { 'Authorization': `Bearer ${token}` } : {}),
      },
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(errorData.message || `Failed to fetch users (status: ${response.status})`);
    }

    return await response.json();
  },

  /**
   * Update the role of a user.
   * @param {string} id - User Guid
   * @param {string} newRole - Target role string
   */
  async updateUserRole(id, newRole) {
    const token = getAuthToken();
    const response = await fetch(`${API_BASE_URL}/users/${id}/role`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
        ...(token ? { 'Authorization': `Bearer ${token}` } : {}),
      },
      body: JSON.stringify({ newRole }),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(errorData.message || `Failed to update user role (status: ${response.status})`);
    }

    return await response.json();
  },
};

export default userService;
